examples/jsm/nodes/shadernode/ShaderNodeUtils.js

```diff
@@ -113,48 +113,57 @@ const ShaderNodeProxy = function ( NodeClass, scope = null, factor = null ) {
 
 		return ( ...params ) => {
 
-			return new ShaderNodeObject( new NodeClass( ...( new ShaderNodeArray( params ) ) ) );
+			return nodeObject( new NodeClass( ...nodeArray( params ) ) );
 
 		};
 
 	} else if ( factor === null ) {
 
 		return ( ...params ) => {
 
-			return new ShaderNodeObject( new NodeClass( scope, ...( new ShaderNodeArray( params ) ) ) );
+			return nodeObject( new NodeClass( scope, ...nodeArray( params ) ) );
 
 		};
 
 	} else {
 
-		factor = new ShaderNodeObject( factor );
+		factor = nodeObject( factor );
 
 		return ( ...params ) => {
 
-			return new ShaderNodeObject( new NodeClass( scope, ...( new ShaderNodeArray( params ) ), factor ) );
+			return nodeObject( new NodeClass( scope, ...nodeArray( params ), factor ) );
 
 		};
 
 	}
 
 };
 
-export const ShaderNodeScript = function ( jsFunc ) {
+const ShaderNodeImmutable = function ( NodeClass, ...params ) {
+
+	return nodeObject( new NodeClass( ...nodeArray( params ) ) );
+
+};
+
+const ShaderNodeScript = function ( jsFunc ) {
 
 	return { call: ( inputs, builder ) => {
 
-		inputs = new ShaderNodeObjects( inputs );
+		inputs = nodeObjects( inputs );
 
-		return new ShaderNodeObject( jsFunc( inputs, builder ) );
+		return nodeObject( jsFunc( inputs, builder ) );
 
 	} };
 
 };
 
-export const nodeObject = ( val ) => new ShaderNodeObject( val );
+export const ShaderNode = new Proxy( ShaderNodeScript, shaderNodeHandler );
```

Comment:
Can you revert `ShaderNode` class to `ShaderNode.js`?

Reply:
I think it is easy to import it from ShaderNodeBaseElements or ShaderNodeElements, I do not see why it should be a separate file...

Reply:
Certainly `ShaderNode`  class shouldn't be in `ShaderNodeUtils.js`, `ShaderNodeUtils` should be a import of `ShaderNode`. New users will search for `ShaderNode.js` file to `ShaderNode` class, all Node System should follow this file terminology. 
In order of easy to import, you can import `ShaderNodeElements` in `ShaderNode` file.

Reply:
Why users should import ShaderNode from `ShaderNode.js` from which they cannot import any of the ShaderNode functions? It would make more sense to import it from somewhere where they can import helper functions (and this is `Nodes.js` or more internal `ShaderNodeElements.js`).

> New users will search for ShaderNode.js file to ShaderNode class, all Node System should follow this file terminology.

I think ShaderNode can be an exception because it does not make sense to import only it without the supplementary functions.

> In order of easy to import, you can import ShaderNodeElements in ShaderNode file.

It is impossible because of the same circular inclusion issue.

Reply:
Maybe if just rename `ShaderNodeUtils` to `ShaderNode` this will make more sense.

Reply:
Done.